common/api/v1.ts

```diff
@@ -25,19 +25,34 @@ export namespace Error {
 export namespace Models {
     /** Absolute position on the page in range [0,1] */
     export class Position {
-        @Validator.validate()
+        @validate()
         x!: number;
-        @Validator.validate()
+        @validate()
         y!: number;
     }
 
-    /** Mazenet user */
+    /** A standarized set of profile information used for authentication
+     */
+    export class Profile {
+        /** provider that the user authenticates with, e.g. email, google, facebook */
+        @validate()
+        provider!: string;
+        /** a unique id identifying the user. in mazenet scheme, this is username
+         * otherwise, it is an id provided by the provider */
+        @validate()
+        id!: string;
+        @validate(true)
+        displayName?: string;
```

Comment:
What is the purpose of this field? Would it become the default username of a new profile unless specified by the user?

Or would this entire class be populated client-side _manually_?

e.g. Client Side Login Flow -> Get oauth JWT client side > populate this class automatically + with user input -> Send this class + JWT to server for registration? 

This is a big PR, maybe I just missed how it this class is used. It seems kinda unnecessary.  We could just send to the server from the client `oauth JWT` + `desired name` and the provider + id could be extracted server-side from the JWT.

Reply:
The display name allows the user to identify what account that they have linked to their mazenet account. The JWT from an identity provider usually comes with several supplementary fields--email, display name, etc. We can store some of this data in the profile, displayName is the most generic, highest level form of this info.

We can allow the user to link multiple accounts to their Mazenet User account, and then use any of those providers to login. Here's how I see it:

 - `User.username`: display name that is show on mazenet to other users. You can change this whenever you want.
 - `Profile.id`: unique id for authentication purposes, cannot be changed. E.g. for the `mazenet` provider this is the username they chose to register their account. For Google OpenID, this is the `sub` claim in the OAuth JWT.
 - `Profile.displayName`: human readable display name identifying the profile. This isn't needed for `mazenet` provider, for Google provider it could be their Full Name, their Email, etc. So if for some reason you attached two separate Google accounts to your mazenet account, you would be able to distinguish them and unlink either one individually.

DisplayName, etc. would be auto-populated by the server when the account is created/linked. We could allow the user to customize the values if they wanted, but it would probably be unnecessary.

Reply:
I see makes sense. I didn't consider attaching multiple Google identities to a single User. I think these systems usually just link as the email being the common connecting key and usually an email is a singleton per identity provider. Gotchu. So you're saying from the client side I shouldn't worry about setting this field at all?

Reply:
You don't need to worry about setting it on the client side.

We could add support to automatically link accounts by email or other pieces of metadata. I designed the system to allow for linking multiple accounts to a single Mazenet User. Whether or not we implement that is a matter of choice--but the choice is there.

Automatically linking the accounts on authentication (rather than through an explicit "link account" flow after the user is logged in) is a potential security concern, as the email can hypothetically be spoofed. For example, you create a Twitter account, set someone else's email as your email, don't "verify" the email, then log into Mazenet with that Twitter account. The server matches the emails and gives you access to someone else's account. Providers usually include in the claims whether the email was verified or not--it's something we need to be careful about and verify for every provider we support.